### PR TITLE
G2gl

### PR DESCRIPTION
**Pull request by mmaiers-nmdp**
added validation for glstrings and fixed a typo

### CODE REVIEW

File: pyard/pyard.py [hunk 1]
```diff
@@ -389,6 +390,9 @@ def redux_gl(self, glstring: str, redux_type: str) -> str:
         :rtype: str
         """
         
+        if not self.isvalid_gl(glstring):
+            return ""
```
Comment (pbashyal-nmdp) [inline pyard/pyard.py]: Should it throw an error rather than silently failing for invalid `glstring`

File: pyard/pyard.py [hunk 2]
```diff
@@ -429,6 +433,42 @@ def redux_gl(self, glstring: str, redux_type: str) -> str:
...
+        """
+        Determines valididy of an allele
+
+        :param allele: An HLA allele.
+        :type: str
+        :return: allele or empty
+        :rtype: boolean
+        """
+        v = lambda a: a in self.valid
+        return v(allele)
```
Comment (pbashyal-nmdp) [inline pyard/pyard.py]:
```suggestion
        return allele in self.valid
```

File: pyard/pyard.py [hunk 3]
```diff
@@ -429,6 +433,42 @@ def redux_gl(self, glstring: str, redux_type: str) -> str:
...
+        Determine validity of glstring
+
+        :param glstring
+        :type: str
+        :return: result
+        :rtype: boolean
+        """
+        
+        if re.search("\^", glstring):
+            return(all(list(map(self.isvalid_gl,glstring.split("^")))))
```
Comment (pbashyal-nmdp) [inline pyard/pyard.py]:
```suggestion
            return(all(map(self.isvalid_gl,glstring.split("^"))))
```
No need to convert it into `list`.

File: pyard/util.py [hunk 4]
```diff
@@ -1,8 +1,8 @@
 # -*- coding: utf-8 -*-
 
 #
-#    seqann Sequence Annotation
-#    Copyright (c) 2017 Be The Match operated by National Marrow Donor Program. All Rights Reserved.
+#    pyard pyARD
+#    Copyright (c) 2018 Be The Match operated by National Marrow Donor Program. All Rights Reserved.
```
Comment (pbashyal-nmdp) [inline pyard/util.py]: 2019?

File: pyard/glstring.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        return "^".join(flatten_loc(g) for g in gls.split("^"))
+    else:
+        return flatten_loc(gls)
+        
+
+def flatten_loc (gls):
+    # if gls contains |
+    if re.search("\|", gls):
+        # loop over all genos
+        typ1 = dict()
```
Comment (pbashyal-nmdp) [inline pyard/glstring.py]: Looks like you're using `dict` just for keys. `set` would probably be better.
```python
type1 = set()
...
type1.add(t1)
...
new_type1 = "/".join(sorted(type1))
```

**Comment by pbashyal-nmdp** [comment]
Where is `glstring` module coming from ? Pypi ? If so it should be included in `requirements.txt` file.

**Comment by mmaiers-nmdp** [comment]
Pradeep,

I like your suggestions but much of that was in the code already.
I’m trying to get this code to Yoram and his team so they can flatten gl strings themselves.
I just realized this pull request has been sitting for a long time.
I’ve got to get slides ready for Navy so I’m not up for fixing style issues in Halagan’s code but you are welcome to…

M

Martin Maiers
Vice President, Biomedical Informatics
CIBMTR® | +1 612 483 9756
mmaiers@nmdp.org<mailto:mmaiers@nmdp.org> | www.cibmtr.org<http://www.cibmtr.org/>
Our goal is to increase event free survival by 25%

**Comment by pbashyal-nmdp** [comment]
Also you're using `flatten` and `flatten_loc` in a loop and using `re.search("\^", gls)` needs to parse regex everytime, you should pre-compile the regex beforehand and use the compiled version to check. eg:
```
gl_split_re = re.compile("\^") # do this once

...
...

if gl_split_re.search(gls):
   ....
```

**Comment by pbashyal-nmdp** [comment]
Ah .. ok. I just got an email for review today.

**Comment by mmaiers-nmdp** [comment]
Yea, there isn’t a button for: just merge this pull request.
But I believe I am the only one in the pyARD code and these are good suggestions so definitely do the review too.

Martin

Martin Maiers
Vice President, Biomedical Informatics
CIBMTR® | +1 612 483 9756
mmaiers@nmdp.org<mailto:mmaiers@nmdp.org> | www.cibmtr.org<http://www.cibmtr.org/>
Our goal is to increase event free survival by 25%

**Pull request merged**